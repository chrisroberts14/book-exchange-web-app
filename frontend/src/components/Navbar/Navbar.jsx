import {
  Navbar,
  NavbarBrand,
  NavbarContent,
  NavbarItem,
  DropdownTrigger,
  Dropdown,
  DropdownMenu,
  DropdownItem,
  Avatar,
  Tabs,
  Tab,
} from "@nextui-org/react";

import { LoginSignUpForm } from "../LoginSignUpForm/LoginSignUpForm.jsx";

export const NavbarComponent = (input) => {
  return (
    <Navbar
      shouldHideOnScroll
      classNames={{
        item: [
          "flex",
          "relative",
          "h-full",
          "items-center",
          "data-[active=true]:after:content-['']",
          "data-[active=true]:after:absolute",
          "data-[active=true]:after:bottom-0",
          "data-[active=true]:after:left-0",
          "data-[active=true]:after:right-0",
          "data-[active=true]:after:h-[3px]",
          "data-[active=true]:after:rounded-[2px]",
          "data-[active=true]:after:bg-primary",
          "data-[active=true]:after:color-primary",
        ],
      }}
      maxWidth={"full"}
    >
      <NavbarBrand href="#">
        <p className="font-bold text-inherit">Book Exchange</p>
      </NavbarBrand>
      <NavbarContent justify="center">
        <NavbarItem>
          <Tabs
            aria-label="Tabs radius"
            className="gap-36"
            color="primary"
            size="lg"
          >
            <Tab key="listings" title="Listings" />
            <Tab key="books" title="Books" />
            <Tab key="users" title="Users" />
          </Tabs>
        </NavbarItem>
      </NavbarContent>
      <NavbarContent justify="end">
        {input.user ? (
          <AvatarDropdown user={input.loggedIn} setUser={input.setUser} />
        ) : (
          <LoginSignUp setUser={input.setUser} />
        )}
      </NavbarContent>
    </Navbar>
  );
};

export const LoginSignUp = (input) => {
  return (
    <Dropdown closeOnSelect={false} placement="bottom-end">
      <DropdownTrigger>
        <Avatar
          isBordered
          as="button"
          className="transition-transform"
          color="secondary"
          name="Account"
          size="sm"
        />
      </DropdownTrigger>
      <DropdownMenu aria-label="Profile Actions">
        <DropdownItem key="sign-in-log-in" isReadOnly>
          <LoginSignUpForm setUser={input.setUser}/>
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
  );
};

export const AvatarDropdown = ({user, setUser}) => {
  return (
    <Dropdown placement="bottom-end">
      <DropdownTrigger data-testid="user-dropdown-open">
        <Avatar
          isBordered
          as="button"
          className="transition-transform"
          color="secondary"
          name={user}
          size="sm"
        />
      </DropdownTrigger>
      <DropdownMenu aria-label="Profile Actions" variant="flat">
        <DropdownItem key="profile" className="h-14 gap-2" textValue="Profile">
          <p className="font-semibold">Signed in as</p>
          <p className="font-semibold">
            {user}
          </p>
        </DropdownItem>
        <DropdownItem key="user_settings" textValue="User Settings">User Settings</DropdownItem>
        <DropdownItem key="user_books" textValue="Your Books">Your Books</DropdownItem>
        <DropdownItem key="user_listings" textValue="Your Listings">Your Listings</DropdownItem>
        <DropdownItem key="logout" color="danger" data-testid="logout-button" onPress={() => {window.localStorage.clear(); setUser(null);}} textValue="Log Out">Log Out</DropdownItem>
      </DropdownMenu>
    </Dropdown>
  );
};
