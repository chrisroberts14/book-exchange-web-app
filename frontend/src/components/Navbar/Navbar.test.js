import {fireEvent, render, screen} from "@testing-library/react";

import {AvatarDropdown, NavbarComponent} from "./Navbar.jsx";

describe("NavbarComponent", () => {
  it("renders correctly logged in", () => {
    const { container } = render(<NavbarComponent signedIn={true} loggedIn={{name: ""}}/>);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly logged out", () => {
    const { container } = render(<NavbarComponent signedIn={false} />);
    expect(container).toMatchSnapshot();
  });
});

describe("Avatar Dropdown", () => {
  let setUser, user;

  beforeEach(() => {
    setUser = jest.fn();
  });

  it("renders correctly", () => {
    const { container } = render(<AvatarDropdown user={user} setUser={setUser} />);
    expect(container).toMatchSnapshot();
  });

  it('logs out correctly', () => {
    render(<AvatarDropdown user={user} setUser={setUser} />);
    fireEvent.click(screen.getByTestId("user-dropdown-open"));
    fireEvent.click(screen.getByTestId("logout-button"));
    expect(setUser).toHaveBeenCalled();
  });
});
