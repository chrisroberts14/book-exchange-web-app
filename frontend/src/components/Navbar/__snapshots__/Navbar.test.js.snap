// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Dropdown renders correctly 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    class="flex relative justify-center items-center box-border overflow-hidden align-middle outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-8 h-8 text-tiny bg-secondary text-secondary-foreground rounded-full ring-2 ring-offset-2 ring-offset-background dark:ring-offset-background-dark ring-secondary z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased transition-transform"
    data-slot="trigger"
    data-testid="user-dropdown-open"
    id="react-aria-:ri:"
    type="button"
  >
    <span
      aria-label="avatar"
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex items-center justify-center text-inherit w-full h-full"
      role="img"
    >
      <svg
        aria-hidden="true"
        fill="none"
        height="80%"
        role="presentation"
        viewBox="0 0 24 24"
        width="80%"
      >
        <path
          d="M12 2C9.38 2 7.25 4.13 7.25 6.75C7.25 9.32 9.26 11.4 11.88 11.49C11.96 11.48 12.04 11.48 12.1 11.49C12.12 11.49 12.13 11.49 12.15 11.49C12.16 11.49 12.16 11.49 12.17 11.49C14.73 11.4 16.74 9.32 16.75 6.75C16.75 4.13 14.62 2 12 2Z"
          fill="currentColor"
        />
        <path
          d="M17.0809 14.1489C14.2909 12.2889 9.74094 12.2889 6.93094 14.1489C5.66094 14.9989 4.96094 16.1489 4.96094 17.3789C4.96094 18.6089 5.66094 19.7489 6.92094 20.5889C8.32094 21.5289 10.1609 21.9989 12.0009 21.9989C13.8409 21.9989 15.6809 21.5289 17.0809 20.5889C18.3409 19.7389 19.0409 18.5989 19.0409 17.3589C19.0309 16.1289 18.3409 14.9889 17.0809 14.1489Z"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`NavbarComponent renders correctly logged in 1`] = `
<div>
  <nav
    class="flex z-40 w-full h-auto items-center justify-center data-[menu-open=true]:border-none sticky top-0 inset-x-0 backdrop-blur-lg data-[menu-open=true]:backdrop-blur-xl backdrop-saturate-150 bg-background/70"
    style="--navbar-height: 4rem; transform: none;"
  >
    <header
      class="z-40 flex px-6 gap-4 w-full flex-row relative flex-nowrap items-center justify-between h-[var(--navbar-height)] max-w-full"
    >
      <div
        class="flex basis-0 flex-row flex-grow flex-nowrap justify-start bg-transparent items-center no-underline text-medium whitespace-nowrap box-border"
        href="#"
      >
        <p
          class="font-bold text-inherit"
        >
          Book Exchange
        </p>
      </div>
      <ul
        class="flex gap-4 h-full flex-row flex-nowrap items-center data-[justify=start]:justify-start data-[justify=start]:flex-grow data-[justify=start]:basis-0 data-[justify=center]:justify-center data-[justify=end]:justify-end data-[justify=end]:flex-grow data-[justify=end]:basis-0"
        data-justify="center"
      >
        <li
          class="text-medium whitespace-nowrap box-border list-none data-[active=true]:font-semibold flex relative h-full items-center data-[active=true]:after:content-[''] data-[active=true]:after:absolute data-[active=true]:after:bottom-0 data-[active=true]:after:left-0 data-[active=true]:after:right-0 data-[active=true]:after:h-[3px] data-[active=true]:after:rounded-[2px] data-[active=true]:after:bg-primary data-[active=true]:after:color-primary"
        >
          <div
            aria-label="Tabs radius"
            class="inline-flex gap-36"
            data-slot="base"
          >
            <div
              aria-label="Tabs radius"
              aria-orientation="horizontal"
              class="flex p-1 h-fit gap-2 items-center flex-nowrap overflow-x-scroll scrollbar-hide bg-default-100 rounded-large"
              data-slot="tabList"
              id="react-aria-:r0:"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="z-0 w-full px-3 py-1 flex group relative justify-center items-center cursor-pointer transition-opacity tap-highlight-transparent data-[disabled=true]:cursor-not-allowed data-[disabled=true]:opacity-30 data-[hover-unselected=true]:opacity-disabled outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-9 text-medium rounded-medium"
                data-key="listings"
                data-selected="true"
                data-slot="tab"
                id="react-aria-:r0:-tab-listings"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="absolute z-0 inset-0 rounded-medium bg-primary text-primary-foreground"
                  data-slot="cursor"
                />
                <div
                  class="relative z-10 whitespace-nowrap transition-colors text-default-500 group-data-[selected=true]:text-primary-foreground"
                  data-slot="tabContent"
                >
                  Listings
                </div>
              </button>
              <button
                aria-selected="false"
                class="z-0 w-full px-3 py-1 flex group relative justify-center items-center cursor-pointer transition-opacity tap-highlight-transparent data-[disabled=true]:cursor-not-allowed data-[disabled=true]:opacity-30 data-[hover-unselected=true]:opacity-disabled outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-9 text-medium rounded-medium"
                data-key="books"
                data-slot="tab"
                id="react-aria-:r0:-tab-books"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <div
                  class="relative z-10 whitespace-nowrap transition-colors text-default-500 group-data-[selected=true]:text-primary-foreground"
                  data-slot="tabContent"
                >
                  Books
                </div>
              </button>
              <button
                aria-selected="false"
                class="z-0 w-full px-3 py-1 flex group relative justify-center items-center cursor-pointer transition-opacity tap-highlight-transparent data-[disabled=true]:cursor-not-allowed data-[disabled=true]:opacity-30 data-[hover-unselected=true]:opacity-disabled outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-9 text-medium rounded-medium"
                data-key="users"
                data-slot="tab"
                id="react-aria-:r0:-tab-users"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <div
                  class="relative z-10 whitespace-nowrap transition-colors text-default-500 group-data-[selected=true]:text-primary-foreground"
                  data-slot="tabContent"
                >
                  Users
                </div>
              </button>
            </div>
          </div>
        </li>
      </ul>
      <ul
        class="flex gap-4 h-full flex-row flex-nowrap items-center data-[justify=start]:justify-start data-[justify=start]:flex-grow data-[justify=start]:basis-0 data-[justify=center]:justify-center data-[justify=end]:justify-end data-[justify=end]:flex-grow data-[justify=end]:basis-0"
        data-justify="end"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="flex relative justify-center items-center box-border overflow-hidden align-middle outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-8 h-8 text-tiny bg-secondary text-secondary-foreground rounded-full ring-2 ring-offset-2 ring-offset-background dark:ring-offset-background-dark ring-secondary z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased transition-transform"
          data-slot="trigger"
          id="react-aria-:r6:"
          type="button"
        >
          <span
            aria-label="Account"
            class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 font-normal text-center text-inherit"
            role="img"
          >
            Acc
          </span>
        </button>
      </ul>
    </header>
  </nav>
</div>
`;

exports[`NavbarComponent renders correctly logged out 1`] = `
<div>
  <nav
    class="flex z-40 w-full h-auto items-center justify-center data-[menu-open=true]:border-none sticky top-0 inset-x-0 backdrop-blur-lg data-[menu-open=true]:backdrop-blur-xl backdrop-saturate-150 bg-background/70"
    style="--navbar-height: 4rem; transform: none;"
  >
    <header
      class="z-40 flex px-6 gap-4 w-full flex-row relative flex-nowrap items-center justify-between h-[var(--navbar-height)] max-w-full"
    >
      <div
        class="flex basis-0 flex-row flex-grow flex-nowrap justify-start bg-transparent items-center no-underline text-medium whitespace-nowrap box-border"
        href="#"
      >
        <p
          class="font-bold text-inherit"
        >
          Book Exchange
        </p>
      </div>
      <ul
        class="flex gap-4 h-full flex-row flex-nowrap items-center data-[justify=start]:justify-start data-[justify=start]:flex-grow data-[justify=start]:basis-0 data-[justify=center]:justify-center data-[justify=end]:justify-end data-[justify=end]:flex-grow data-[justify=end]:basis-0"
        data-justify="center"
      >
        <li
          class="text-medium whitespace-nowrap box-border list-none data-[active=true]:font-semibold flex relative h-full items-center data-[active=true]:after:content-[''] data-[active=true]:after:absolute data-[active=true]:after:bottom-0 data-[active=true]:after:left-0 data-[active=true]:after:right-0 data-[active=true]:after:h-[3px] data-[active=true]:after:rounded-[2px] data-[active=true]:after:bg-primary data-[active=true]:after:color-primary"
        >
          <div
            aria-label="Tabs radius"
            class="inline-flex gap-36"
            data-slot="base"
          >
            <div
              aria-label="Tabs radius"
              aria-orientation="horizontal"
              class="flex p-1 h-fit gap-2 items-center flex-nowrap overflow-x-scroll scrollbar-hide bg-default-100 rounded-large"
              data-slot="tabList"
              id="react-aria-:r9:"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="z-0 w-full px-3 py-1 flex group relative justify-center items-center cursor-pointer transition-opacity tap-highlight-transparent data-[disabled=true]:cursor-not-allowed data-[disabled=true]:opacity-30 data-[hover-unselected=true]:opacity-disabled outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-9 text-medium rounded-medium"
                data-key="listings"
                data-selected="true"
                data-slot="tab"
                id="react-aria-:r9:-tab-listings"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="absolute z-0 inset-0 rounded-medium bg-primary text-primary-foreground"
                  data-slot="cursor"
                />
                <div
                  class="relative z-10 whitespace-nowrap transition-colors text-default-500 group-data-[selected=true]:text-primary-foreground"
                  data-slot="tabContent"
                >
                  Listings
                </div>
              </button>
              <button
                aria-selected="false"
                class="z-0 w-full px-3 py-1 flex group relative justify-center items-center cursor-pointer transition-opacity tap-highlight-transparent data-[disabled=true]:cursor-not-allowed data-[disabled=true]:opacity-30 data-[hover-unselected=true]:opacity-disabled outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-9 text-medium rounded-medium"
                data-key="books"
                data-slot="tab"
                id="react-aria-:r9:-tab-books"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <div
                  class="relative z-10 whitespace-nowrap transition-colors text-default-500 group-data-[selected=true]:text-primary-foreground"
                  data-slot="tabContent"
                >
                  Books
                </div>
              </button>
              <button
                aria-selected="false"
                class="z-0 w-full px-3 py-1 flex group relative justify-center items-center cursor-pointer transition-opacity tap-highlight-transparent data-[disabled=true]:cursor-not-allowed data-[disabled=true]:opacity-30 data-[hover-unselected=true]:opacity-disabled outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-9 text-medium rounded-medium"
                data-key="users"
                data-slot="tab"
                id="react-aria-:r9:-tab-users"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <div
                  class="relative z-10 whitespace-nowrap transition-colors text-default-500 group-data-[selected=true]:text-primary-foreground"
                  data-slot="tabContent"
                >
                  Users
                </div>
              </button>
            </div>
          </div>
        </li>
      </ul>
      <ul
        class="flex gap-4 h-full flex-row flex-nowrap items-center data-[justify=start]:justify-start data-[justify=start]:flex-grow data-[justify=start]:basis-0 data-[justify=center]:justify-center data-[justify=end]:justify-end data-[justify=end]:flex-grow data-[justify=end]:basis-0"
        data-justify="end"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="flex relative justify-center items-center box-border overflow-hidden align-middle outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-8 h-8 text-tiny bg-secondary text-secondary-foreground rounded-full ring-2 ring-offset-2 ring-offset-background dark:ring-offset-background-dark ring-secondary z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased transition-transform"
          data-slot="trigger"
          id="react-aria-:rf:"
          type="button"
        >
          <span
            aria-label="Account"
            class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 font-normal text-center text-inherit"
            role="img"
          >
            Acc
          </span>
        </button>
      </ul>
    </header>
  </nav>
</div>
`;
